paperparcel-kotlin/src/main/java/paperparcel/PaperParcelable.kt

```diff
@@ -11,38 +11,38 @@ import java.lang.reflect.Method
  * @since 2.0
  */
 interface PaperParcelable : Parcelable {
-  private object Cache {
-    val writeMethods = mutableMapOf<Class<*>, Method>()
-  }
-
   override fun describeContents() = 0
 
   override fun writeToParcel(dest: Parcel, flags: Int) {
-    if (!Cache.writeMethods.containsKey(javaClass)) {
-      val annotatedClass = findAnnotatedClass(javaClass)
-      val generatedClass = Class.forName(implementationName(annotatedClass))
-      val args = arrayOf(annotatedClass, Parcel::class.java, Int::class.javaPrimitiveType)
-      val writeMethod = generatedClass.getDeclaredMethod("writeToParcel", *args)
-      writeMethod.isAccessible = true
-      Cache.writeMethods.put(javaClass, writeMethod)
+    val writeMethod: Method = synchronized(writeMethods) {
+      writeMethods.getOrPut(javaClass) {
+        val annotatedClass = javaClass.annotatedClass
+        val generatedClass = Class.forName(annotatedClass.implementationName)
+        generatedClass
+                .getDeclaredMethod("writeToParcel", annotatedClass, Parcel::class.java, Int::class.javaPrimitiveType)
+                .apply {
+                  isAccessible = true
+                }
+      }
     }
-    Cache.writeMethods[javaClass]!!.invoke(null, this, dest, flags)
+    writeMethod.invoke(null, this, dest, flags)
```

Comment:
This is placed outside of the synchronised block since it could lead to deadlocks otherwise.

Reply:
This code looks much nicer than what I had, thanks!

I would change the formatting to something more like this though: 
```
Class.forName(annotatedClass.implementationName)
    .getDeclaredMethod("writeToParcel", annotatedClass, Parcel::class.java, Int::class.javaPrimitiveType)
    .apply { isAccessible = true }
```